File: app/src/main/java/com/fafu/app/dfb/mvp/login/LoginModel.java
package com.fafu.app.dfb.mvp.login;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Base64;

import com.fafu.app.dfb.data.UserMe;
import com.fafu.app.dfb.http.LoginService;
import com.fafu.app.dfb.http.RetrofitFactory;
import com.fafu.app.dfb.mvp.base.BaseModel;
import com.fafu.app.dfb.util.RxJavaUtils;
import com.fafu.app.dfb.util.SPUtils;

import java.io.IOException;

import io.reactivex.Observable;
import okhttp3.ResponseBody;

public class LoginModel extends BaseModel implements LoginContract.Model {

    private final LoginService service = RetrofitFactory.obtainService(LoginService.class, null);
    private boolean isInit = false;

    @Override
    public UserMe getUserMe() {
        if (SPUtils.get("UserInfo").contain("sno") && SPUtils.get("UserInfo").contain("password")) {
            UserMe user = new UserMe();
            user.setSno(SPUtils.get("UserInfo").getString("sno"));
            user.setPassword(SPUtils.get("UserInfo").getString("password"));
            return user;
        }
        return null;
    }

    @Override
    public Observable<Bitmap> verifyBitmap() {
        return RxJavaUtils.create( emitter -> {
            if (!isInit) {
                isInit = true;
                service.init("0", SPUtils.get("Const").getString("IMEI"), "0").execute();
            }
            ResponseBody responseBody = service.verify(
                    String.valueOf(System.currentTimeMillis())
            ).execute().body();
            try {
                assert responseBody != null;
                byte[] bytes = responseBody.bytes();
                emitter.onNext(BitmapFactory.decodeByteArray(bytes, 0, bytes.length));
            } catch (IOException e) {
                emitter.onError(e);
            }
            emitter.onComplete();
        });
    }

    @Override
    public Observable<String> login(String sno, String password, String verify) {
        return RxJavaUtils.create( emitter -> {
            ResponseBody responseBody = service.login(
                    "http://cardapp.fafu.edu.cn:8088/Phone/Login?sourcetype=0&IMEI=" + SPUtils.get("Const").getString("IMEI") + "&language=0",
                    sno, new String(Base64.encode(password.getBytes(), Base64.DEFAULT)),
                    verify, "1", "1", "", "true"
            ).execute().body();
            assert responseBody != null;
            emitter.onNext(responseBody.string());
            emitter.onComplete();
        });
    }

    @Override
    public void save(UserMe user, String rescouseType) {
        SPUtils.get("UserInfo").putString("account", user.getAccount());
        SPUtils.get("UserInfo").putString("sno", user.getSno());
        SPUtils.get("UserInfo").putString("password", user.getPassword());
        SPUtils.get("UserInfo").putString("name", user.getName());
        SPUtils.get("Cookie").putString("RescouseType", rescouseType);
    }

}

File: app/src/main/java/com/fafu/app/dfb/mvp/login/LoginModel.kt
package com.fafu.app.dfb.mvp.login

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.fafu.app.dfb.data.UserMe
import com.fafu.app.dfb.http.RetrofitFactory
import com.fafu.app.dfb.http.XFBaoService
import com.fafu.app.dfb.util.SPUtils
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody

class LoginModel {

    private val service = RetrofitFactory.obtainService(XFBaoService::class.java)
    private val TAG = "LoginModel"

    fun verify(): Observable<Bitmap> {
        return Observable
                .create<Bitmap> {
                    val resp = service.verify().execute()
                    convert(resp.body())?.run {
                        it.onNext(this)
                    }
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun login(account: String, password: String, verify: String): Observable<String> {
        return Observable
                .create<String> {
                    val resp = service.login(
                            referer = "http://cardapp.fafu.edu.cn:8088/Phone/Login?sourcetype=0&IMEI=${com.fafu.app.dfb.util.StringUtils.imei()}&language=0",
                            sno = account,
                            pwd = String(Base64.encode(password.toByteArray(), Base64.DEFAULT)),
                            yzm = verify
                    ).execute().body()
                    if (resp != null) {
                        it.onNext(resp.string())
                    }
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun save(user: UserMe) {
        SPUtils.putString("sno", user.sno)
        SPUtils.putString("password", user.password)
        SPUtils.putString("name", user.name)
        SPUtils.putString("RescouseType", user.rescouseType)
    }

    private fun convert(value: ResponseBody?): Bitmap? {
        val bytes = value?.bytes() ?: return null
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }
}